src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1271,16 +1281,23 @@ private void Abort(Exception abortException)
 
         private void AbortStreams(int lastValidStream, Exception abortException)
         {
-            bool shouldInvalidate = false;
-            lock (SyncObject)
+            Debug.Assert(lastValidStream >= 0);
+            bool hasAnyActiveStream = false;
+
+            if (_lastStreamId == -1)
```

Comment:
I think this logic should be under the lock, to synchronize properly with the logic in AddStream.

Reply:
Aren't the frames processed synchronously? Why is the lock is needed?

Reply:
Yes, frames are processed synchronously, but AddStream happens when a request is sent and can happen concurrently with this logic.

Reply:
I'll move it under the lock since it's only few instructions and it's better be safe than sorry but still curious why is lock even needed

Reply:
Looking at the code, I think it's probably safe to not have this under the lock. But it's not immediately obvious why. So if you want to keep it outside the lock, you need to add a comment explaining exactly why this is safe so it doesn't get inadvertantly broken in the future.

My preference would be to just put it under the lock so it's clearly correct, and we don't risk breaking this in the future. There doesn't really seem to be much reason not to put it under the lock.